sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java

```diff
@@ -334,94 +336,88 @@ public void setup() {
     public void teardown() {
       jedis.close();
     }
-  }
-
-  private static class ReadKeysWithPattern extends BaseReadFn<String> {
 
-    ReadKeysWithPattern(RedisConnectionConfiguration connectionConfiguration) {
-      super(connectionConfiguration);
+    @GetInitialRestriction
+    public OffsetRange getInitialRestriction(@Element String pattern) {
+      return new OffsetRange(0, getKeyPatters(pattern).size());
     }
 
     @ProcessElement
-    public void processElement(ProcessContext c) {
-      ScanParams scanParams = new ScanParams();
-      scanParams.match(c.element());
-
-      String cursor = ScanParams.SCAN_POINTER_START;
-      boolean finished = false;
-      while (!finished) {
-        ScanResult<String> scanResult = jedis.scan(cursor, scanParams);
-        List<String> keys = scanResult.getResult();
-        for (String k : keys) {
-          c.output(k);
+    public void processElement(ProcessContext c, RestrictionTracker<OffsetRange, Long> tracker) {
+      String pattern = c.element();
+      List<String> keys = getKeyPatters(pattern);
+      List<String> bundle = new ArrayList<>();
+      for (long i = tracker.currentRestriction().getFrom();
+          i < tracker.currentRestriction().getTo();
+          i++) {
+        if (tracker.tryClaim(i)) {
+          bundle.add(keys.get((int) i));
         }
-        cursor = scanResult.getCursor();
-        if (cursor.equals(ScanParams.SCAN_POINTER_START)) {
-          finished = true;
+      }
+      if (bundle.size() > 0) {
+        List<KV<String, String>> kvs = fetchAndFlush(bundle);
```

Comment:
Probably loading the whole bundle to memory here could result in OOMs.

Reply:
Updated